pkg/rlptools/wasm_utils.go

```diff
@@ -10,119 +10,193 @@ import (
-		{
-			Actions: []kuadrantv1beta1.ActionSpecifier{
-				{
-					GenericKey: &kuadrantv1beta1.GenericKeySpec{
-						DescriptorValue: key.String(),
-						// using default value as specified in Envoy spec
-						// https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/route/v3/route_components.proto#config-route-v3-ratelimit-action-generickey
-						DescriptorKey: &[]string{"ratelimitpolicy"}[0],
-					},
-				},
-			},
-		},
+func ruleFromLimit(limitFullName string, limit *kuadrantv1beta2.Limit, route *gatewayapiv1alpha2.HTTPRoute) wasm.Rule {
+	if limit == nil {
+		return wasm.Rule{}
+	}
+
+	return wasm.Rule{
+		Conditions: conditionsFromLimit(limit, route),
+		Data:       dataFromLimt(limitFullName, limit),
 	}
 }
 
-// GatewayActionsFromRateLimitPolicy return flatten list from GatewayAction from the RLP
-func GatewayActionsFromRateLimitPolicy(rlp *kuadrantv1beta1.RateLimitPolicy, route *gatewayapiv1alpha2.HTTPRoute) []GatewayAction {
-	flattenActions := make([]GatewayAction, 0)
-	if rlp == nil {
-		return flattenActions
+func conditionsFromLimit(limit *kuadrantv1beta2.Limit, route *gatewayapiv1alpha2.HTTPRoute) []wasm.Condition {
+	if limit == nil {
+		return make([]wasm.Condition, 0)
 	}
 
-	for idx := range rlp.Spec.RateLimits {
-		// Skip those RateLimit objects with empty configurations, even if they have rules defined
-		if len(rlp.Spec.RateLimits[idx].Configurations) == 0 {
-			continue
+	// TODO(eastizle): review this implementation. This is a first naive implementation.
+	// The conditions must always be a subset of the route's matching rules.
+
+	conditions := make([]wasm.Condition, 0)
+
+	for routeSelectorIdx := range limit.RouteSelectors {
+		// TODO(eastizle): what if there are only Hostnames (i.e. empty "matches" list)
+		for matchIdx := range limit.RouteSelectors[routeSelectorIdx].Matches {
+			condition := wasm.Condition{
+				AllOf: patternExpresionsFromMatch(&limit.RouteSelectors[routeSelectorIdx].Matches[matchIdx]),
+			}
+
+			// merge hostnames expression in the same condition
+			for _, hostname := range limit.RouteSelectors[routeSelectorIdx].Hostnames {
+				condition.AllOf = append(condition.AllOf, patternExpresionFromHostname(hostname))
+			}
+
+			conditions = append(conditions, condition)
 		}
-		// if HTTPRoute is available, fill empty rules with defaults from the route
-		rules := rlp.Spec.RateLimits[idx].Rules
-		if route != nil && len(rules) == 0 {
-			rules = HTTPRouteRulesToRLPRules(common.RulesFromHTTPRoute(route))
+	}
+
+	if len(conditions) == 0 {
+		conditions = append(conditions, conditionsFromRoute(route)...)
+	}
+
+	// merge when expression in the same condition
+	// must be done after adding route level conditions when no route selector are available
+	// prevent conditions only filled with "when" definitions
+	for whenIdx := range limit.When {
+		for idx := range conditions {
+			conditions[idx].AllOf = append(conditions[idx].AllOf, patternExpresionFromWhen(limit.When[whenIdx]))
 		}
+	}
```

Comment:
We're drifting a bit from the proposed "soft conditions" implemented in Limitador (see last bullet point at https://github.com/Kuadrant/architecture/blob/main/rfcs/0001-rlp-v2.md#highlights), by going one step ahead to improve performance.

By evaluating the RLP's `when` condition as a normal condition in the Wasm shim (just like the conditions built from `routeSelectors`), the RL request will only be sent to Limitador when there's truly at least one limit whose counters to increment.